src/sentry/tasks/summaries/daily_summary.py

```diff
@@ -207,12 +207,41 @@ def build_summary_data(
                 group__in=([group for group in regressed_or_escalated_groups]),
                 type__in=(ActivityType.SET_REGRESSION.value, ActivityType.SET_ESCALATING.value),
             )
-            if regressed_or_escalated_groups_today:
-                for activity in regressed_or_escalated_groups_today[:4]:
-                    if activity.type == ActivityType.SET_REGRESSION.value:
-                        project_ctx.regressed_today.append(activity.group)
-                    else:
-                        project_ctx.escalated_today.append(activity.group)
+
+            deduped_groups_by_activity_type: dict[int, set] = {
+                ActivityType.SET_REGRESSION.value: set(),
+                ActivityType.SET_ESCALATING.value: set(),
+            }
+            for activity in regressed_or_escalated_groups_today:
+                if activity.type == ActivityType.SET_ESCALATING.value:
+                    # if a group is already in the regressed set but we now see it in escalating, remove from regressed and add to escalating
+                    # this means the group regressed and then later escalated, and we only want to list it once
+                    if (
+                        activity.group
+                        in deduped_groups_by_activity_type[ActivityType.SET_REGRESSION.value]
+                    ):
+                        deduped_groups_by_activity_type[ActivityType.SET_REGRESSION.value].remove(
+                            activity.group
+                        )
+                    deduped_groups_by_activity_type[ActivityType.SET_ESCALATING.value].add(
+                        activity.group
+                    )
+                elif (
+                    activity.type == ActivityType.SET_REGRESSION.value
+                    and activity.group
+                    not in deduped_groups_by_activity_type[ActivityType.SET_ESCALATING.value]
+                ):
+                    deduped_groups_by_activity_type[ActivityType.SET_REGRESSION.value].add(
+                        activity.group
+                    )
+
+            for activity_type, groups in deduped_groups_by_activity_type.items():
+                if activity_type == ActivityType.SET_REGRESSION.value:
+                    for group in list(groups)[:4]:
```

Comment:
Could you move this loop / logic up a level since it's being repeated in both `if` and `else`?